### PR TITLE
xds: allow for peered upstreams to use tagged addresses that are hostnames

### PR DESCRIPTION
**Pull request by rboyer**
### Description

Mesh gateways [can use hostnames in their tagged addresses](https://github.com/hashicorp/consul/pull/7999/files). This is useful if you were to expose a mesh gateway using a cloud networking load balancer appliance that gives you a DNS name but no reliable static IPs.

Envoy cannot accept hostnames via EDS and those must be configured using CDS. There was already logic when configuring gateways in other locations in the code, but given the illusions in play for peering the downstream of a peered service wasn't aware that it should be doing that.

Also:
- ensuring that we always try to use wan-like addresses to cross peer boundaries.

### Links

- [Asana - wan hostnames](https://app.asana.com/0/1201832349181086/1202428211194184/f)
- [Asana - wan addr](https://app.asana.com/0/0/1202428211194183/f)

### CODE REVIEW

File: agent/proxycfg/testing_peering.go [hunk 1]
```diff
@@ -51,6 +51,16 @@ func TestConfigSnapshotPeering(t testing.T) *ConfigSnapshot {
 							Service: "payments-sidecar-proxy",
```
Comment (rboyer) [inline agent/proxycfg/testing_peering.go:51]: There are two cross-peer upstreams here and I've only adjusted one of them so we can test both scenarios at once.

File: agent/xds/endpoints.go [hunk 2]
```diff
@@ -96,14 +96,20 @@ func (s *ResourceGenerator) endpointsFromSnapshotConnectProxy(cfgSnap *proxycfg.
...
+
 		endpoints, ok := cfgSnap.ConnectProxy.PeerUpstreamEndpoints[uid]
 		if ok {
 			la := makeLoadAssignment(
 				clusterName,
 				[]loadAssignmentEndpointGroup{
 					{Endpoints: endpoints},
 				},
-				cfgSnap.Locality,
+				GatewayKey{ /*empty so it never matches*/ },
```
Comment (rboyer) [inline agent/xds/endpoints.go:112]: For the EDS cases, this ensures we use wan-like addresses.

File: agent/xds/clusters.go [hunk 3]
```diff
@@ -602,6 +596,35 @@ func (s *ResourceGenerator) makeUpstreamClusterForPeerService(
...
+
+		useEDS := true
+		if _, ok := cfgSnap.ConnectProxy.PeerUpstreamEndpointsUseHostnames[uid]; ok {
+			useEDS = false
+		}
+
+		// If none of the service instances are addressed by a hostname we
+		// provide the endpoint IP addresses via EDS
+		if useEDS {
+			c.ClusterDiscoveryType = &envoy_cluster_v3.Cluster_Type{Type: envoy_cluster_v3.Cluster_EDS}
```
Comment (kisunji) [inline agent/xds/clusters.go:608]: nit: this is already set in L583

**Review by kisunji — APPROVED** [review]

File: agent/xds/clusters.go [hunk 4]
```diff
@@ -1054,17 +1077,39 @@ func (s *ResourceGenerator) makeGatewayCluster(snap *proxycfg.ConfigSnapshot, op
...
+	return cluster
+}
+
+func configureClusterWithHostnames(
+	logger hclog.Logger,
+	cluster *envoy_cluster_v3.Cluster,
+	dnsDiscoveryType string,
+	// hostnameEndpoints is a list of endpoints with a hostname as their address
+	hostnameEndpoints structs.CheckServiceNodes,
+	// isRemote determines whether the cluster is in a remote DC and we should prefer a WAN address
```
Comment (freddygv) [inline agent/xds/clusters.go:1100]:
```suggestion
	// isRemote determines whether the cluster is in a remote DC or partition and we should prefer a WAN address
```

**Review by freddygv — APPROVED** [review]
LGTM, one minor suggestion

**Pull request merged**